internal/state/backend_refs.go

```diff
@@ -0,0 +1,152 @@
+package state
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/state/resolver"
+)
+
+type (
+	// ruleIndex is the index of the HTTPRouteRule.
+	ruleIndex int
+	// resolvedBackends is a map of backend names to their endpoints.
+	resolvedBackends map[string][]resolver.Endpoint
+	// backendGroupsByRule is a map of rule indexes to backend groups.
+	backendGroupsByRule map[ruleIndex]BackendGroup
+)
+
+// BackendRefs includes the BackendRefs of an HTTPRoute.
+type BackendRefs struct {
+	Errors   []string
+	Resolved resolvedBackends
+	ByRule   backendGroupsByRule
+}
+
+func newBackendRefs() BackendRefs {
+	return BackendRefs{
+		Errors:   make([]string, 0),
+		Resolved: make(resolvedBackends),
+		ByRule:   make(backendGroupsByRule),
+	}
+}
+
+func resolveBackendRefs(
+	ctx context.Context,
+	routes map[types.NamespacedName]*route,
+	services map[types.NamespacedName]*v1.Service,
+	resolver resolver.ServiceResolver,
+) Warnings {
+	resolveBackendRefsForRoutes(ctx, routes, services, resolver)
+
+	warnings := newWarnings()
+	for _, r := range routes {
+		for _, msg := range r.BackendRefs.Errors {
+			warnings.AddWarningf(
+				r.Source,
+				"cannot resolve backend ref: %s",
+				msg,
+			)
+		}
+	}
+
+	return warnings
+}
+
+func resolveBackendRefsForRoutes(
+	ctx context.Context,
+	routes map[types.NamespacedName]*route,
+	services map[types.NamespacedName]*v1.Service,
+	resolver resolver.ServiceResolver,
+) {
+	for _, r := range routes {
+		for idx, rule := range r.Source.Spec.Rules {
+
+			backends := make([]BackendRef, 0, len(rule.BackendRefs))
+
+			for _, ref := range rule.BackendRefs {
+
+				weight := int32(1)
+				if ref.Weight != nil {
+					weight = *ref.Weight
+				}
+
+				svc, port, err := getServiceAndPortFromRef(ref.BackendRef, r.Source.Namespace, services)
+				if err != nil {
+					backends = append(backends, BackendRef{Weight: weight})
+
+					r.BackendRefs.Errors = append(r.BackendRefs.Errors, err.Error())
+
+					continue
+				}
+
+				backendName := fmt.Sprintf("%s_%s_%d", svc.Namespace, svc.Name, port)
+
+				backends = append(backends, BackendRef{
+					Name:   backendName,
+					Valid:  true,
+					Weight: weight,
+				})
+
+				eps, err := resolver.Resolve(ctx, svc, int32(*ref.Port))
```

Comment:
Potentially we have this problem across rules as well. Same Service, same Route, different Rules resolve at different times. One has different Endpoints than the other(s).

Is this temporary? Presumably the change to the cache is accompanied by a new reconciliation event which rectifies the difference. Will we become 'eventually' consistent in these cases or get stuck? I'm not sure how to test this deterministically.

Reply:
> Considering that, would it make sense to perform deduplication before we start resolving backends?
Additionally, because resolver.Resolve calls the cache directly, the cache can change from call to call (since its updated async), so in the same route we can get a resolved service and a service resolved with an error.

@pleshakov I spoke offline with @f5yacobucci and we came to the conclusion that since this issue is temporary -- we will quickly get a reconcile event that will trigger a reconfig of nginx -- it's not necessary that we fix it. The system is behaving as an eventually consistent system should behave.